src/index.ts

```diff
@@ -1,387 +1,431 @@
+
+
+    }
+
+    /**
+     * getFullHostName
+     * Just generates the full connection name for amqp based on the passed in config
+     * @param config - Contains the amqp connection config
+     **/
+    private static getFullHostName(config: IAmqpCacoonConfig) {
+        var fullHostNameString =
+            config.protocol +
+            '://' +
+            config.username +
+            ':' +
+            config.password +
+            '@' +
+            config.host;
+
+        if (config.port) {
+            fullHostNameString = fullHostNameString + ':' + config.port;
+        }
+        return fullHostNameString;
     }
 
-    return fullHostNameString;
-  }
-
-  /**
-   * getPublishChannel
-   * This connects to amqp and creates a channel or gets the current channel
-   * @return channel
-   **/
-  async getPublishChannel() {
-    try {
-      // Return the pubChannel if we are already connected
-      if (this.pubChannel) return this.pubChannel;
-      // Connect if needed
-      this.connection =
-        this.connection ||
-        (await amqp.connect(this.fullHostName, this.amqp_opts));
-      // Open a channel
-      this.pubChannel = await this.connection.createChannel();
-    } catch (e) {
-      if (this.logger) this.logger.error('AMQPCacoon.connect: Error: ', e);
-      throw e;
+    private injectConnectionEvents(connection: AmqpConnectionManager) {
+        // Subscribe to onConnect / onDisconnection functions for debugging
+        connection.on('connect', () => {
+            this.handleBrokerConnect();
+        });
+
+        connection.on('disconnect', () => {
+            this.handleBrokerDisonnect();
+        });
     }
-    // Return the channel
-    return this.pubChannel;
-  }
-
-  /**
-   * getConsumerChannel
-   * This connects to amqp and creates a channel or gets the current channel
-   * @return channel
-   **/
-  async getConsumerChannel() {
-    try {
-      // Return the subChannel if we are already connected
-      if (this.subChannel) return this.subChannel;
-      // Connect if needed
-      this.connection =
-        this.connection ||
-        (await amqp.connect(this.fullHostName, this.amqp_opts));
-      // Open a channel
-      this.subChannel = await this.connection.createChannel();
-    } catch (e) {
-      if (this.logger)
-        this.logger.error('AMQPCacoon.getConsumerChannel: Error: ', e);
-      throw e;
+
+    /**
+     * getPublishChannel
+     * This connects to amqp and creates a channel or gets the current channel
+     * @return ChannelWrapper
+     **/
+    async getPublishChannel() {
+        try {
+            // Return the pubChannel if we are already connected
+            if (this.pubChannelWrapper) {
+                return this.pubChannelWrapper;
+            }
+            // Connect if needed
+            this.connection =
+                this.connection ||
+                (await amqp.connect([this.fullHostName], this.amqp_opts));
```

Comment:
Interesting, it appears that amqp.connect doesn't return a promise. We should be able to remove the await here.

Reply:
If you think I'm wrong then feel free to ignore this comment because it doesn't actually hurt the code.

Reply:
Yeah, you're correct!

Reply:
updated